parser/parser_test.go

```diff
@@ -639,23 +639,276 @@ func TestParseValidInput(t *testing.T) {
+	for _, aKey := range aKeys {
+		if !(isIn(aKey, bKeys)) {
+			aValue := a.MapIndex(aKey)
+			return false, fmt.Sprintf("%v contains key %s (with value %#v) but %v does not", actual.name, aKey, aValue, expected.name)
+		}
+	}
+	for _, bKey := range bKeys {
+		if !(isIn(bKey, aKeys)) {
+			bValue := b.MapIndex(bKey)
+			return false, fmt.Sprintf("%v contains key %s (with value %#v) but %v does not", expected.name, bKey, bValue, actual.name)
+		}
+	}
+	for _, aKey := range aKeys {
+		bKey, _ := findMatching(aKey, bKeys)
+		aIndexText := fmt.Sprintf("[%v(%t)]", aKey, bKey)
+		bIndexText := fmt.Sprintf("[%v(%t)]", bKey, bKey)
+
+		equal, diff = deepCompareObjects(namedItem{actual.name + aIndexText, a.MapIndex(aKey)},
+			namedItem{expected.name + bIndexText, b.MapIndex(bKey)},
+			depth,
+		)
+		if !equal {
+			return equal, diff
+		}
+
+	}
+	return true, ""
+}
+
+func deepCompareArrays(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	a := actual.item
+	b := expected.item
+	aLen := a.Len()
+	bLen := b.Len()
+	sharedLen := int(math.Min(float64(aLen), float64(bLen)))
+	for i := 0; i < sharedLen; i++ {
+		indexText := fmt.Sprintf("[%v]", i)
+		equal, diff = deepCompareObjects(namedItem{actual.name + indexText, a.Index(i)},
+			namedItem{expected.name + indexText, b.Index(i)},
+			depth,
+		)
+		if !equal {
+			return equal, diff
+		}
+	}
+	if aLen != bLen {
+		var mismatchIndex int
+		var mismatchName string
+		var mismatchValue interface{}
+		if aLen < bLen {
+			mismatchIndex = aLen
+			mismatchName = expected.name
+			mismatchValue = expected.item.Index(mismatchIndex)
 		} else {
-			t.Errorf("test case: %s", testCase.input)
-			t.Errorf("length mismatch, expected %d comments, got %d",
-				len(testCase.comments), len(file.Comments))
+			mismatchIndex = bLen
+			mismatchName = actual.name
+			mismatchValue = actual.item.Index(mismatchIndex)
+		}
+		return false, fmt.Sprintf("%s.len() = %v whereas %s.len() = %#v.\n\nFirst differing item: %s[%v] = %s",
+			actual.name, aLen, expected.name, bLen, mismatchName, mismatchIndex, mismatchValue)
+	}
+	return true, ""
+}
+
+func deepCompareStructs(actual namedItem, expected namedItem, depth int) (equal bool, diff string) {
+	a := actual.item
+	b := expected.item
+	aCount := a.NumField()
+	bCount := b.NumField()
+	if aCount != bCount {
+		panic(fmt.Sprintf("Internal error: Objects of different classes were sent to deepCompareStructs. It should be caught by deepCompareObjects higher on the stack. "+
+			"Objects: %#v, %#v", a, b))
+	}
+	aType := a.Type()
+	for i := 0; i < aCount; i++ {
+		fieldName := aType.Field(i).Name
+		equal, diff = deepCompareObjects(namedItem{actual.name + "." + fieldName, a.Field(i)},
+			namedItem{expected.name + "." + fieldName, b.Field(i)},
+			depth)
+		if !equal {
+			return equal, diff
 		}
 	}
+	return true, ""
+}
+
+// deepCompare is much like reflect.DeepEqual, but it does a deep comparison of map keys too (which is slower than a map lookup)
```

Comment:
Yeah, I was surprised too. I'll add a test involving a map having keys of pointer type.